geomesa-gs-security/src/main/scala/org/geomesa/gs/security/audit/ElasticRequestDataListener.scala

```diff
@@ -0,0 +1,119 @@
+/***********************************************************************
+ * Copyright (c) 2013-2020 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the GNU GENERAL PUBLIC LICENSE,
+ * Version 2 which accompanies this distribution and is available at
+ * https://opensource.org/licenses/GPL-2.0.
+ ***********************************************************************/
+
+package org.geomesa.gs.security.audit
+
+import java.lang.reflect.Type
+import java.util.Date
+
+import com.google.gson._
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.http.HttpResponse
+import org.apache.http.client.config.RequestConfig
+import org.apache.http.client.methods.{HttpPost}
+import org.apache.http.concurrent.FutureCallback
+import org.apache.http.entity.StringEntity
+import org.apache.http.impl.nio.client.{CloseableHttpAsyncClient, HttpAsyncClients}
+import org.geoserver.monitor.{RequestData, RequestDataListener}
+import org.opengis.geometry.BoundingBox
+
+import scala.concurrent.Await
+
+
+class ElasticRequestDataListener extends RequestDataListener with LazyLogging {
+
+  private val requestConfig = RequestConfig.custom()
+    .setSocketTimeout(10000)
+    .setConnectTimeout(10000).build();
+
+  val client: CloseableHttpAsyncClient = HttpAsyncClients.custom()
+    .setMaxConnTotal(10)
+    .setDefaultRequestConfig(requestConfig)
+    .build()
+
+  client.start()
+  val host = sys.env.getOrElse("ELASTICSEARCH_HOST", null)
+  val index = sys.env.getOrElse("GEOSERVER_ES_INDEX", null)
+
+  private val gson: Gson = new GsonBuilder()
+    .registerTypeAdapter(classOf[Date], new DateSerializer)
+    .registerTypeAdapter(classOf[BoundingBox], new BoundingBoxSerializer)
+    .registerTypeAdapter(classOf[Throwable], new ThrowableSerializer)
+    .serializeNulls().create()
+
+  class DateSerializer extends JsonSerializer[Date] {
+    override def serialize(src: Date, typeOfSrc: Type, context: JsonSerializationContext): JsonElement = {
+      new JsonPrimitive(src.getTime)
+    }
+  }
+
+   class BoundingBoxSerializer extends JsonSerializer[BoundingBox] {
+     override def serialize(src: BoundingBox, typeOfSrc: Type, context: JsonSerializationContext): JsonElement = {
+       new JsonPrimitive(src.toString)
+     }
+   }
+
+  class ThrowableSerializer extends JsonSerializer[Throwable] {
+    override def serialize(src: Throwable, typeOfSrc: Type, context: JsonSerializationContext): JsonElement = {
+      new JsonPrimitive(src.getMessage)
+    }
+  }
```

Comment:
i'd suggest moving these to a companion object since the can be shared across instances.

Reply:
@elahrvivaz I'm unfamiliar with Scala so I'm not sure what that would look like. I understand what a companion object is supposed to be in simple terms just struggling to apply it here.

Reply:
oh sure, it's pretty easy. you just add an `object` with the same name after the class definition, like:
```
class ElasticRequestDataListener {
  import ElasticRequestDataListener.gson
  gson.toJson(???)
}

object ElasticRequestDataListener {
   private val gson: Gson = ???
}
```

The things in the class are tied to a class instance, like with java inner classes. The things in the object are singletons shared across all instances of the class, like java static values. The class can also access private vals in the companion object. So for things that don't require any access to the instance class vals/defs, like the json serialization classes, we can move them to the object. The gson val is also thread-safe, so we can move it to the companion object and share a single one.

In this particular case, it probably matters less, since there will only ever be a single instance of a listener (I assume), but i think it makes the code easier to understand and more organized.

Reply:
ah okay great, thank you